q2_demux/tests/test_filter.py

```diff
@@ -0,0 +1,245 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2019, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import unittest
+import os
+
+import pandas as pd
+
+from qiime2 import Metadata
+from qiime2.plugin.testing import TestPluginBase
+from q2_types.per_sample_sequences import (
+    SingleLanePerSampleSingleEndFastqDirFmt,
+    SingleLanePerSamplePairedEndFastqDirFmt)
+
+from q2_demux._filter import filter_samples
+from q2_demux._summarize import _PlotQualView
+
+
+class FilterSamplesTests(TestPluginBase):
+    package = 'q2_demux.tests'
+
+    def setUp(self):
+        super().setUp()
+
+        data_single = SingleLanePerSampleSingleEndFastqDirFmt(
+            self.get_data_path('filter_samples_single_end/dir_fmt'), mode='r')
+        self.sample_single = _PlotQualView(data_single, False)
+        self.manifest_single = data_single.manifest.view(pd.DataFrame)
+
+        self.md_single_all = Metadata.load(
+            self.get_data_path('filter_samples_single_end/filter_all.tsv'))
+        self.md_single_subset = Metadata.load(
+            self.get_data_path('filter_samples_single_end/filter_subset.tsv'))
+        self.md_single_none = Metadata.load(
+            self.get_data_path('filter_samples_single_end/filter_none.tsv'))
+
+        data_paired = SingleLanePerSamplePairedEndFastqDirFmt(
+            self.get_data_path('filter_samples_paired_end/dir_fmt'), mode='r')
+        self.sample_paired = _PlotQualView(data_paired, True)
+        self.manifest_paired = data_paired.manifest.view(pd.DataFrame)
+
+        self.md_paired_all = Metadata.load(
+            self.get_data_path('filter_samples_single_end/filter_all.tsv'))
+        self.md_paired_subset = Metadata.load(
+            self.get_data_path('filter_samples_single_end/filter_subset.tsv'))
+        self.md_paired_none = Metadata.load(
+            self.get_data_path('filter_samples_single_end/filter_none.tsv'))
+
+    def _assert(self, obs_ids, exp_ids):
+        self.assertEqual(set(obs_ids), set(exp_ids))
+
+    def _extract_single(self, obs_files):
+        return [file.split('_', 1)[0] for file in os.listdir(str(obs_files))]
+
+    def _extract_paired(self, obs_files):
+        obs_ids = [(file.split('_')[0::3]) for file in
+                   os.listdir(str(obs_files))]
+        return [(id[0] + id[1]) for id in obs_ids]
```

Comment:
You could merge these three functions into two:

```
def assert_single_contains(dir_fmt, exp_ids):
    ...

def assert_paired_contains(dir_fmt, exp_ids):
    ...
```

creating custom assertions that do a bit of work to unpack/inspect a structure is a very common thing in our code-bases.

Reply:
to clarify, the three functions are the ones immediately above this comment position: `_assert`, `_extract_single`, and `_extract_paired`